semantic_router/llms/openai.py

```diff
@@ -36,22 +37,70 @@ def __init__(
         self.temperature = temperature
         self.max_tokens = max_tokens
 
-    def __call__(self, messages: List[Message]) -> str:
+    def _extract_tool_calls_info(self, tool_calls: list[dict[str, Any]]) -> list[dict[str, Any]]:
+        tool_calls_info = []
+        for tool_call in tool_calls:
+            if tool_call.function.arguments is None:
+                raise ValueError(
+                    "Invalid output, expected arguments to be specified for each tool call."
+                )
+            tool_calls_info.append({
+                "function_name": tool_call.function.name,
+                "arguments": tool_call.function.arguments
+            })
+        return json.dumps(tool_calls_info)
+    
+    def __call__(
+        self,
+        messages: List[Message],
+        function_schemas: Optional[list[dict[str, Any]]] = None,
+    ) -> str:
         if self.client is None:
             raise ValueError("OpenAI client is not initialized.")
         try:
+            if function_schemas:
+                tools = function_schemas
+            else:
+                tools = None
+
             completion = self.client.chat.completions.create(
                 model=self.name,
                 messages=[m.to_openai() for m in messages],
                 temperature=self.temperature,
                 max_tokens=self.max_tokens,
+                tools=tools,  # type: ignore # MyPy expecting Iterable[ChatCompletionToolParam] | NotGiven, but dict is accepted by OpenAI.
             )
 
-            output = completion.choices[0].message.content
+            if function_schemas:
+                tool_calls = completion.choices[0].message.tool_calls
+                if tool_calls is None:
+                    raise ValueError("Invalid output, expected a tool call.")
+                if len(tool_calls) < 1:
+                    raise ValueError(
+                        "Invalid output, expected at least one tool to be specified."
+                    )
+                
+                # Collecting multiple tool calls information
+                output = self._extract_tool_calls_info(tool_calls)
+            else:
+                content = completion.choices[0].message.content
+                if content is None:
+                    raise ValueError("Invalid output, expected content.")
+                output = str(content)  # str to keep MyPy happy.
```

Comment:
Is this because `content` can also be `None`? If so shouldn't that cause an error to be raised?

Reply:
It was something like that. Basically mypy wasn't clever enough to understand that the logic meant it was impossible for `output` to be anything but `str` type. 

I had to wrap content in `str()` to make it stop complaining.

However since then the logic has changed and I can remove `str()` without mypy complaining. So will do that.